core/compiler.common/src/org/jetbrains/kotlin/types/SmartcastStability.kt

```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.types
+
+enum class SmartcastStability(private val str: String, val description: String = str) {
+    // Local value, or parameter, or private / internal member value without open / custom getter,
+    // or protected / public member value from the same module without open / custom getter
+    // Smart casts are completely safe
+    STABLE_VALUE("stable val"),
+
+    // Member value with open / custom getter
+    // Smart casts are not safe
+    PROPERTY_WITH_GETTER("custom getter", "property that has open or custom getter"),
+
+    // Protected / public member value from another module
+    // Smart casts are not safe
+    ALIEN_PUBLIC_PROPERTY("alien public", "public API property declared in different module"),
+
+    // Local variable already captured by a changing closure
+    // Smart casts are not safe
+    CAPTURED_VARIABLE("captured var", "local variable that is captured by a changing closure"),
+
+    // Member variable regardless of its visibility
+    // Smart casts are not safe
+    MUTABLE_PROPERTY("member", "mutable property that could have been changed by this time"),
+
+    DELEGATED_PROPERTY("delegate", "delegated property that could have been changed by this time"),
+}
```

Comment:
Yeah, it's a decision that need to be made. So as for now non-property expressions won't be treated as smartcast at all. I am not sure if we want to do this or mimic FE1.0 to add smartcast to any expression. Consider the following

```
fun testSmartcast(f: Foo) {
    if (f.getSomething() is String) {
        f.getSomething().length
    }
}
```

FE1.0 would highlight `f.getSomething()` in the if body as smartcast impossible due to complex expression. With the current FIR implementation, it seems to be non-trivial to consider non-property access as smartcasts. @demiurg906 what do you think?

Reply:
I think we should not treat such expressions as potentially smartcasted expressions. Unstable smartcasts on properties is OK, because property may represent stable reference to single object. Functions has ho guaranties about that by definition, so smartcast (even non possible) on it is nonsence